include/mbedtls/mbedtls_config.h

```diff
@@ -2672,6 +2672,25 @@
  */
 #define MBEDTLS_MD_C
 
+/**
+ * \def MBEDTLS_MD_LIGHT
+ *
+ * Enable the "light" subset of MBEDTLS_MD_C: just hashing and basic
+ * meta-data.
+ *
+ * This is automatically enabled whenever MBEDTLS_MD_C is enabled, but it is
+ * possible to enable this with MBEDTLS_MD_C if support for HMAC or extra
+ * metadata functions is not needed.
+ *
+ * Requires: one of: MBEDTLS_MD5_C, MBEDTLS_RIPEMD160_C, MBEDTLS_SHA1_C,
+ *                   MBEDTLS_SHA224_C, MBEDTLS_SHA256_C, MBEDTLS_SHA384_C,
+ *                   MBEDTLS_SHA512_C.
+ * Module:  library/md.c
+ *
+ * Uncomment to enabled the "light" subsect of MD.
+ */
+#define MBEDTLS_MD_LIGHT
```

Comment:
Shouldn't this be named `MBEDTLS_MD_LIGHT_C`?

Reply:
I don't know, I think both make sense. There's no file `md_light.c` which is why I went without `_C`. OTOH, we have names with `_C` for things that don't correspond to the name of a file, like `SHA224_C` and `SHA384_C` - but there I think it's for uniformity with the other `SHAxxx_C`. Here it's a bit of a special case: unlike the various `SHAxxx_C` that are now independent of each other, here `MD_LIGHT` is explicitly a subset of `MD_C` and the later auto-enables the former.

Unless you and/or the other reviewer feel strongly about this, I'm inclined to keep the current name just to avoid changes.

Reply:
The historical convention is `_C` for features that are implemented by a separate `.c` file. But that doesn't make a lot of sense since the fact that the implementation is a separate `.c` file is irrelevant to the programming interface. And so `MBEDTLS_SHA224_C` and `MBEDTLS_SHA384_C` departed from this historical convention because as an interface, it makes sense to make them similar to the other hashes.

PSA and TLS have been deviating from this convention since we refactored the code to split it into more files, but didn't make an API change just for that. `{md,pk,cipher}_wrap.c` also deviate: they don't have their own enablement macro because they aren't useful on their own.

Usually `MBEDTLS_XXX_YYY` is a subset of `MBEDTLS_XXX_C` which either has no effect if `MBEDTLS_XXX_C` is disabled, or requires `MBEDTLS_XXX_C` to be enabled. This is an interface consideration which `MBEDTLS_MD_LIGHT` would deviate from. We have at least one precedent: `MBEDTLS_PSA_CRYPTO_CLIENT` is a subset of `MBEDTLS_PSA_CRYPTO_C`.

When looking at md alone, `MBEDTLS_MD_LIGHT` seems right to me. When looking at the bigger picture, `MBEDTLS_MD_LIGHT_C` (considering it to be a separate module, which `MBEDTLS_MD_C` happens to require) is more uniform. But `MBEDTLS_MD_LIGHT` is essentially an internal symbol, so I don't think interface uniformity matters much.

In fact, _should_ we even expose `MBEDTLS_MD_LIGHT` in `mbedtls_config.h`, as opposed to just automatically enabling it when other modules need it? Do we really want to promise this as a stable interface?

Reply:
> In fact, _should_ we even expose `MBEDTLS_MD_LIGHT` in `mbedtls_config.h`, as opposed to just automatically enabling it when other modules need it? Do we really want to promise this as a stable interface?

Indeed initially I was planning on keeping it internal (that's what I did in the prototype). Then I thought it might be useful for people who just need hashing but not HMAC, they can save a bit of code size this way. (Actually you can use all of X.509, as well as AEAD ciphersuites for TLS 1.2 without HMAC, so that might be a number of users.)

Also, it feels a bit simpler for all build options to be in `mbedtls_config.h` (though that's already not the case for some options, like those related to driver testing).

Wdyt?

Reply:
I'm a bit wary of introducing a new interface and declaring it as stable straight away. In this case, it's constrained to be a subset of the existing MD interface, and we've already pretty much reduced it to the minimum, so we're unlikely to make incompatible changes. But maybe we'll want to make some optimizations that break the API, e.g. changing the behavior on invalid arguments?

Reply:
I'd prefer to make it internal to start with, if only to reduce the potential testing burden. Then once the dust has settled, maybe expose it — depending partly on whether the dust does settle quickly.

Reply:
Sounds like a plan. Let's do that.